ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py

```diff
@@ -0,0 +1,187 @@
+                            )
+                        )
+                        self.towers[-1].create_ppo_optimizer()
+
+            # Create reward signals
+            for reward_signal, config in reward_signal_configs.items():
+                self.reward_signals[reward_signal] = create_reward_signal(
+                    self, reward_signal, config
+                )
+
+            # Create pretrainer if needed
+            if "pretraining" in trainer_params:
+                BCModule.check_config(trainer_params["pretraining"])
+                self.bc_module = BCModule(
+                    self,
+                    policy_learning_rate=trainer_params["learning_rate"],
+                    default_batch_size=trainer_params["batch_size"],
+                    default_num_epoch=trainer_params["num_epoch"],
+                    **trainer_params["pretraining"],
+                )
+            else:
+                self.bc_module = None
+
+        self.model = self.towers[0]
+
+        if load:
+            self._load_graph()
+        else:
+            self._initialize_graph()
+
+        self.inference_dict = {
+            "action": self.model.output,
+            "log_probs": self.model.all_log_probs,
+            "value": self.model.value_heads,
+            "entropy": self.model.entropy,
+            "learning_rate": self.model.learning_rate,
+        }
+        if self.use_continuous_act:
+            self.inference_dict["pre_action"] = self.model.output_pre
+        if self.use_recurrent:
+            self.inference_dict["memory_out"] = self.model.memory_out
+        if (
+            is_training
+            and self.use_vec_obs
+            and trainer_params["normalize"]
+            and not load
+        ):
+            self.inference_dict["update_mean"] = self.model.update_normalization
+
+        self.update_batch = self.average_gradients([t.grads for t in self.towers])
+        self.update_dict = {"update_batch": self.update_batch}
+        self.update_dict.update({"value_loss_%d".format(i):self.towers[i].value_loss for i in range(len(self.towers)) })
+        self.update_dict.update({"policy_loss_%d".format(i):self.towers[i].policy_loss for i in range(len(self.towers)) })
+
+
+    @timed
+    def update(self, mini_batch, n_sequences):
+        """
+        Updates model using buffer.
+        :param n_sequences: Number of trajectories in batch.
+        :param mini_batch: Experience batch.
+        :return: Output from update process.
+        """
+        feed_dict = {}
+        towers = self.model.towers if len(self.devices) > 1 else [self.model]
+
+        device_batch_size = n_sequences // len(self.devices)
+        device_batches = []
+        for i in range(len(self.devices)):
+            device_batches.append(
+                {k: v[i : i + device_batch_size] for k, v in mini_batch.items()}
+            )
+
+        assert len(device_batches) == len(towers)
+        for batch, tower in zip(device_batches, towers):
```

Comment:
It seems that the logic below for constructing the feed dict is common between `PPOPolicy` and `MultiGPUPPOPolicy`.  I'd recommend you factor the logic into a function like `PPOPolicy.construct_feed_dict(model, minibatch, num_sequences)` and then use that here to gather the per-graph components of the feed dict so we don't duplicate logic.